File: controller/api/sso.py
"""The endpoints behind a "Sign in with…" button, and the screen that sets it up.

Three of these are reachable without a session, which is unavoidable -- they
are how a session is obtained -- so each one is written to be safe for a
stranger to call:

* `/api/auth/providers` returns a name and a button and nothing else. Not the
  client id, not the issuer, not whether anybody has ever signed in with it.
* `/api/auth/sso/{id}/start` only redirects. It creates a row that expires in
  ten minutes and reveals nothing about the studio.
* `/api/auth/sso/{id}/callback` refuses anything whose `state` it did not
  itself write, which is the whole of CSRF protection for this flow.

Everything else here needs an administrator.
"""
from __future__ import annotations

import json
import time
from urllib.parse import quote

from fastapi import APIRouter, Body, HTTPException, Query, Request
from fastapi.responses import RedirectResponse

from .. import auth, config, db, directory, oidc
from .accounts import _set_cookie
from .security import require_admin

router = APIRouter(prefix="/api")


def public_base(request: Request) -> str:
    """The address a browser reaches this studio at.

    Not `request.base_url`, at least not first. A redirect URI has to match the
    one registered with the provider *exactly*, and behind any reverse proxy
    the URL the application sees is the internal one -- http, port 8420,
    container hostname -- which matches nothing anybody registered.
    """
    if fixed := config.PUBLIC_URL:
        return fixed.rstrip("/")
    proto = request.headers.get("x-forwarded-proto", "").split(",")[0].strip()
    host = request.headers.get("x-forwarded-host", "").split(",")[0].strip()
    if proto and host:
        return "%s://%s" % (proto, host)
    return str(request.base_url).rstrip("/")


def redirect_uri(request: Request, idp_id: str) -> str:
    return "%s/api/auth/sso/%s/callback" % (public_base(request), idp_id)


# ---------------------------------------------------------------------------
# Signing in
# ---------------------------------------------------------------------------

@router.get("/auth/providers")
async def providers() -> dict:
    idps = [oidc.public_idp(i) for i in db.list_idps(enabled_only=True)]
    return {
        "providers": idps,
        # Password sign-in is never taken away by configuring SSO. A studio
        # whose provider is unreachable at 8am must not be a studio nobody can
        # get into, and somebody always has to be able to fix it.
        "password_login": True,
    }


@router.get("/auth/sso/{idp_id}/start")
async def start(request: Request, idp_id: str, next: str | None = None):
    idp = db.get_idp(idp_id)
    if not idp or not idp["enabled"]:
        return _back(request, "That sign-in method is not available here.")
    try:
        url = oidc.begin(idp, redirect_uri(request, idp_id), _safe_next(next))
    except oidc.SsoError as e:
        return _back(request, str(e))
    return RedirectResponse(url, status_code=302)


@router.get("/auth/sso/{idp_id}/callback")
async def callback(request: Request, idp_id: str,
                   code: str | None = None, state: str | None = None,
                   error: str | None = None,
                   error_description: str | None = None):
    if error:
        # The provider itself refused -- consent declined, app disabled, user
        # not assigned. Its words are more useful than any of ours.
        return _back(request, error_description or error)
    if not code or not state:
        return _back(request, "That sign-in did not complete.")
    try:
        user, next_url = await oidc.complete(state, code)
    except oidc.SsoError as e:
        return _back(request, str(e))

    response = RedirectResponse(
        public_base(request) + "/" + (next_url or ""), status_code=302)
    _set_cookie(response, request,
                auth.start_session(user["id"],
                                   request.headers.get("user-agent", "")))
    auth.purge_expired_sessions()
    return response


def _safe_next(value: str | None) -> str | None:
    """Only ever a hash route on this studio.

    An open redirect is the classic way a sign-in flow becomes a phishing
    tool: the link looks like ours, and the destination is not.
    """
    value = (value or "").strip()
    if value.startswith("#/") and "\\" not in value and "//" not in value[1:]:
        return value
    return None


def _back(request: Request, message: str) -> RedirectResponse:
    return RedirectResponse(
        "%s/?sso_error=%s" % (public_base(request), quote(message[:400])),
        status_code=302)


# ---------------------------------------------------------------------------
# Setting it up
# ---------------------------------------------------------------------------

def free_slug(kind: str) -> str:
    """The id a provider of this kind would get.

    Providers are named after their kind rather than given a random id, and it
    matters: the redirect URI contains the id, and it has to be registered at
    the provider *before* this studio has one to give. A predictable slug is
    what lets the setup screen show the exact address to paste while the form
    is still empty.
    """
    taken = {i["id"] for i in db.list_idps()}
    if kind not in taken:
        return kind
    for n in range(2, 50):
        if "%s%d" % (kind, n) not in taken:
            return "%s%d" % (kind, n)
    raise HTTPException(400, "That is a lot of sign-in methods.")


@router.get("/idp/presets")
async def presets(request: Request) -> dict:
    require_admin(request)
    base = public_base(request)
    return {
        "presets": oidc.PRESETS,
        "base": base,
        # What each kind would be called, and therefore where its provider
        # should be told to send people back to.
        "redirect_uris": {kind: "%s/api/auth/sso/%s/callback"
                                % (base, free_slug(kind))
                          for kind in oidc.PRESETS},
    }


@router.get("/idp")
async def list_idps(request: Request) -> list[dict]:
    require_admin(request)
    out = []
    for idp in db.list_idps():
        row = oidc.admin_idp(idp)
        row["redirect_uri"] = redirect_uri(request, idp["id"])
        out.append(row)
    return out


def _settings(payload: dict, existing: dict | None = None) -> dict:
    """Validate what an administrator typed, once, for create and for edit."""
    fields: dict = {}
    if "name" in payload:
        fields["name"] = (payload.get("name") or "").strip()[:60]
    for flag in ("enabled", "auto_create", "link_by_email", "sync_enabled"):
        if flag in payload:
            fields[flag] = 1 if payload[flag] else 0
    for text in ("allowed_domains", "admin_groups", "sync_group",
                 "sync_subject", "scopes"):
        if text in payload:
            fields[text] = (payload.get(text) or "").strip()[:400] or None
    if "client_id" in payload:
        client_id = (payload.get("client_id") or "").strip()
        if not client_id:
            raise HTTPException(400, "The application (client) ID is required.")
        fields["client_id"] = client_id
    if payload.get("client_secret"):
        fields["client_secret_enc"] = auth.encrypt_secret(payload["client_secret"])
    elif payload.get("client_secret") == "":
        fields["client_secret_enc"] = None
    if payload.get("sync_secret"):
        fields["sync_secret_enc"] = auth.encrypt_secret(payload["sync_secret"])
    elif payload.get("sync_secret") == "":
        fields["sync_secret_enc"] = None

    # Turning on sync without a way to read the directory is the mistake worth
    # catching here rather than at 2am when the timer first fires.
    kind = (existing or {}).get("kind") or payload.get("kind")
    if fields.get("sync_enabled") and not directory.supports({"kind": kind}):
        raise HTTPException(
            400, "There is no standard way to read a list of people out of an "
                 "OpenID Connect provider. Accounts here are created as people "
                 "sign in instead.")
    return fields


@router.post("/idp")
async def create_idp(request: Request, payload: dict = Body(...)) -> dict:
    admin = require_admin(request)
    kind = (payload.get("kind") or "oidc").strip()
    if kind not in oidc.PRESETS:
        raise HTTPException(400, "Unknown kind of provider.")
    try:
        issuer = oidc.issuer_for(kind, payload.get("tenant") or "")
        doc = await oidc.discover(issuer)
    except oidc.SsoError as e:
        raise HTTPException(400, str(e)) from e

    fields = _settings(payload, {"kind": kind})
    if not fields.get("client_id"):
        raise HTTPException(400, "The application (client) ID is required.")
    fields["name"] = fields.get("name") or oidc.PRESETS[kind]["label"]
    idp_id = db.create_idp(
        free_slug(kind), kind=kind, issuer=issuer,
        tenant=(payload.get("tenant") or "").strip() or None,
        discovery=json.dumps(doc), discovered_at=time.time(),
        created_by=admin["id"], **fields)
    if not db.get_idp(idp_id)["scopes"]:
        db.update_idp(idp_id, scopes=oidc.PRESETS[kind]["scopes"])
    row = oidc.admin_idp(db.get_idp(idp_id))
    row["redirect_uri"] = redirect_uri(request, idp_id)
    return row


@router.patch("/idp/{idp_id}")
async def modify_idp(request: Request, idp_id: str,
                     payload: dict = Body(...)) -> dict:
    require_admin(request)
    idp = db.get_idp(idp_id)
    if not idp:
        raise HTTPException(404, "No such sign-in method.")
    fields = _settings(payload, idp)
    if "tenant" in payload:
        try:
            issuer = oidc.issuer_for(idp["kind"], payload["tenant"])
            fields["issuer"] = issuer
            fields["tenant"] = (payload["tenant"] or "").strip() or None
        except oidc.SsoError as e:
            raise HTTPException(400, str(e)) from e
    db.update_idp(idp_id, **fields)
    if fields.get("issuer"):
        await _rediscover(idp_id)
    row = oidc.admin_idp(db.get_idp(idp_id))
    row["redirect_uri"] = redirect_uri(request, idp_id)
    return row


async def _rediscover(idp_id: str) -> dict:
    idp = db.get_idp(idp_id)
    doc = await oidc.discover(idp["issuer"])
    db.update_idp(idp_id, discovery=json.dumps(doc), discovered_at=time.time())
    return doc


@router.post("/idp/{idp_id}/rediscover")
async def rediscover(request: Request, idp_id: str) -> dict:
    require_admin(request)
    if not db.get_idp(idp_id):
        raise HTTPException(404, "No such sign-in method.")
    try:
        await _rediscover(idp_id)
    except oidc.SsoError as e:
        raise HTTPException(400, str(e)) from e
    return oidc.admin_idp(db.get_idp(idp_id))


@router.delete("/idp/{idp_id}")
async def remove_idp(request: Request, idp_id: str) -> dict:
    require_admin(request)
    if not db.get_idp(idp_id):
        raise HTTPException(404, "No such sign-in method.")
    detached = db.delete_idp(idp_id)
    return {"ok": True, "detached": detached,
            "note": "%d account(s) kept their runs but can no longer sign in "
                    "this way. Give them a password to let them back in."
                    % detached}


@router.post("/idp/{idp_id}/sync")
async def run_sync(request: Request, idp_id: str,
                   dry_run: bool = Query(False)) -> dict:
    require_admin(request)
    idp = db.get_idp(idp_id)
    if not idp:
        raise HTTPException(404, "No such sign-in method.")
    try:
        return await directory.sync(idp, dry_run=dry_run)
    except directory.SyncError as e:
        raise HTTPException(400, str(e)) from e

File: web/js/views/sso.js
/**
 * Connecting the studio to the place your accounts already live.
 *
 * Setting up OpenID Connect is a two-sided job — half of it happens at the
 * provider, in a console this page cannot see — and every failure mode is a
 * mismatch between the two halves. So this screen is built around that: the
 * redirect URI is shown *before* anything is saved, in its final form, ready
 * to paste; discovery runs on save and reports what the provider actually
 * published; and a directory sync can be run as a dry run first, which reads
 * everything and writes nothing.
 *
 * Secrets go in and never come back out. The form shows whether one is set,
 * never what it is, and leaving the field empty on an edit keeps the one
 * already stored rather than clearing it.
 */
import { api } from "../api.js";
import { html, raw, esc, $, $$, on, toast, modal, fmtAgo } from "../util.js";
import { ribbon, rb, group } from "../ribbon.js";
import { breadcrumb } from "../components.js";

export async function ssoView(mount) {
  let idps = [];
  let presets = {};

  const paint = async () => {
    [idps, presets] = await Promise.all([api.idps(), api.idpPresets()]);
    mount.innerHTML = layout(idps, presets);
  };
  await paint();

  on(mount, "click", "#addIdp", () => chooseKind(presets, paint));
  on(mount, "click", "[data-edit]", (_e, t) =>
    openForm(idps.find((i) => i.id === t.dataset.edit), presets, paint));

  on(mount, "click", "[data-toggle]", async (_e, t) => {
    try {
      await api.updateIdp(t.dataset.toggle, { enabled: t.dataset.to === "1" });
      await paint();
    } catch (e) { toast(e.message, "err"); }
  });

  on(mount, "click", "[data-rediscover]", async (_e, t) => {
    t.disabled = true;
    try {
      await api.rediscoverIdp(t.dataset.rediscover);
      toast("Read the provider's configuration again.", "ok");
      await paint();
    } catch (e) { toast(e.message, "err"); t.disabled = false; }
  });

  on(mount, "click", "[data-sync]", async (_e, t) => {
    const dry = t.dataset.dry === "1";
    t.disabled = true;
    t.textContent = dry ? "Reading…" : "Syncing…";
    try {
      const r = await api.syncIdp(t.dataset.sync, dry);
      toast(dry
        ? `Read ${r.read} people. ${r.created} would be new. Nothing changed.`
        : `${r.read} read · ${r.created} added · ${r.updated} updated · `
          + `${r.deactivated} disabled.`, r.refused ? "err" : "ok");
      await paint();
    } catch (e) { toast(e.message, "err"); await paint(); }
  });

  on(mount, "click", "[data-del-idp]", async (_e, t) => {
    const idp = idps.find((i) => i.id === t.dataset.delIdp);
    if (!confirm(`Remove ${idp.name}?\n\n${idp.people} account(s) came from `
                 + `it. They keep everything they own, but they will not be `
                 + `able to sign in until somebody gives them a password.`)) return;
    try {
      const r = await api.deleteIdp(idp.id);
      toast(r.note, "ok");
      await paint();
    } catch (e) { toast(e.message, "err"); }
  });

  on(mount, "click", "[data-copy]", (_e, t) => {
    navigator.clipboard.writeText(t.dataset.copy)
      .then(() => toast("Copied.", "ok"))
      .catch(() => toast("Could not copy.", "err"));
  });
}

// ---------------------------------------------------------------------------

function layout(idps, presets) {
  return html`
    <div class="page-head">
      ${raw(breadcrumb({ href: "#/settings", label: "Settings" }))}
      <h1 style="margin:6px 0 0">Single sign-on</h1>
      <p class="sub">Let people in with the account they already have, and
        find colleagues by name when sharing — without keeping a second list
        of who works here.</p>
    </div>
    ${raw(ribbon({
      tabs: [{ key: "home", label: "Sign-in methods" }], active: "home",
      body: group("Methods", [
        rb("addIdp", "＋", "Add a method", { cls: "primary" }),
      ]) + group("Elsewhere", [
        rb(null, "◍", "People", { href: "#/users" }),
        rb(null, "⚙", "Settings", { href: "#/settings" }),
      ]),
    }))}

    ${raw(!idps.length ? empty() : idps.map(card).join(""))}

    <div class="callout" style="margin-top:16px">
      <strong>Passwords keep working.</strong>
      Connecting a provider never takes the password form away. A studio whose
      identity provider is unreachable at eight in the morning must not be a
      studio nobody can get into — and somebody has to be able to fix it.
    </div>`;
}

const empty = () => html`
  <div class="card">
    <div class="empty" style="padding:28px 16px">
      <div class="big">\u{1F511}</div>
      <h3>Nobody signs in from outside yet</h3>
      <p class="muted" style="max-width:52ch;margin:0 auto">
        Connect Entra ID, Google Workspace, Okta, Keycloak or anything else
        that speaks OpenID Connect. It takes two values from your provider \u2014
        an application ID and a secret \u2014 and one address pasted back the
        other way.</p>
    </div>
  </div>`;

function card(idp) {
  const ok = idp.endpoints?.authorization_endpoint;
  return html`
    <div class="card" style="margin-bottom:14px">
      <div class="row-between" style="gap:10px;flex-wrap:wrap;align-items:flex-start">
        <div style="min-width:0">
          <h3 style="margin-bottom:2px">${idp.name}
            ${raw(idp.enabled
              ? `<span class="badge badge-ok">on</span>`
              : `<span class="badge">off</span>`)}
          </h3>
          <p class="muted tiny mono" style="margin:0;word-break:break-all">${idp.issuer}</p>
        </div>
        <div class="row" style="gap:6px;flex-wrap:wrap">
          <button class="btn-sm" data-edit="${idp.id}">Settings</button>
          <button class="btn-sm" data-toggle="${idp.id}"
                  data-to="${idp.enabled ? "0" : "1"}">
            ${idp.enabled ? "Turn off" : "Turn on"}</button>
          <button class="btn-sm btn-danger" data-del-idp="${idp.id}">Remove</button>
        </div>
      </div>

      <div class="factrow" style="margin-top:12px">
        <div class="fact"><div class="k">People</div>
          <div class="v">${idp.people}</div>
          <div class="n">${idp.pending} not signed in yet</div></div>
        <div class="fact"><div class="k">New accounts</div>
          <div class="v">${idp.auto_create ? "Created" : "Refused"}</div>
          <div class="n">${idp.allowed_domains
            ? idp.allowed_domains : "any address"}</div></div>
        <div class="fact"><div class="k">Directory</div>
          <div class="v">${idp.sync_enabled ? "Syncing" : (idp.directory_kind ? "Off" : "n/a")}</div>
          <div class="n">${idp.last_sync_at ? fmtAgo(idp.last_sync_at)
            : (idp.directory_kind || "not available")}</div></div>
        <div class="fact"><div class="k">Configuration</div>
          <div class="v">${ok ? "Read" : "Missing"}</div>
          <div class="n">${idp.discovered_at ? fmtAgo(idp.discovered_at) : "—"}</div></div>
      </div>

      <div class="field" style="margin-top:6px">
        <label>Redirect URI — this exact string goes in the provider</label>
        <div class="row row-top">
          <input readonly value="${idp.redirect_uri}">
          <button class="btn-sm" data-copy="${idp.redirect_uri}">Copy</button>
        </div>
        <div class="hint">Providers compare it character for character. A
          mismatch is the single most common reason a sign-in fails, and it
          always fails at the provider, before this studio hears anything.</div>
      </div>

      ${raw(idp.last_sync_note ? html`
        <p class="muted tiny" style="margin:8px 0 0">
          Last sync: ${idp.last_sync_note}</p>` : "")}

      <div class="row" style="gap:6px;margin-top:10px;flex-wrap:wrap">
        <button class="btn-sm" data-rediscover="${idp.id}">Re-read configuration</button>
        ${raw(idp.directory_kind ? html`
          <button class="btn-sm" data-sync="${idp.id}" data-dry="1">Preview a sync</button>
          <button class="btn-sm btn-primary" data-sync="${idp.id}">Sync people now</button>`
        : html`<span class="muted tiny" style="align-self:center">
            This kind of provider has no list of people to read, so accounts
            here appear as people sign in.</span>`)}
      </div>
    </div>`;
}

// ---------------------------------------------------------------------------
// Adding one
// ---------------------------------------------------------------------------

function chooseKind(presets, refresh) {
  const dlg = modal({ title: "What are you connecting?", width: 560 });
  $(".modal-body", dlg).innerHTML = html`
    <div class="picklist">
      ${raw(Object.entries(presets.presets || {}).map(([kind, p]) => html`
        <button class="kind-option" data-kind="${kind}">
          <strong>${p.label}</strong>
          <span class="muted tiny">${p.blurb}</span>
          ${raw(p.directory ? `<span class="badge badge-soft">reads your
            directory</span>` : "")}
        </button>`).join(""))}
    </div>`;
  on(dlg, "click", "[data-kind]", (_e, t) => {
    dlg.close();
    openForm(null, presets, refresh, t.dataset.kind);
  });
}

function openForm(existing, presets, refresh, kind) {
  kind = kind || existing?.kind || "oidc";
  const preset = (presets.presets || {})[kind] || {};
  const redirect = existing?.redirect_uri
    || (presets.redirect_uris || {})[kind] || "";
  const dlg = modal({
    title: existing ? `${existing.name} settings` : `Connect ${preset.label}`,
    width: 640,
  });

  $(".modal-body", dlg).innerHTML = html`
    <div class="callout" style="margin-bottom:14px">
      <strong>First, at ${preset.label}.</strong>
      Register an application and give it this redirect URI:
      <div class="row row-top" style="margin-top:8px">
        <input readonly value="${redirect}">
        <button class="btn-sm" type="button" data-copy-redirect>Copy</button>
      </div>
      ${raw(preset.docs ? html`<div class="hint">${preset.docs}</div>` : "")}
    </div>

    <form id="idpForm">
      <div class="field">
        <label for="f_name">Name on the sign-in button</label>
        <input id="f_name" name="name" value="${existing?.name || preset.label || ""}"
               placeholder="${preset.label || "Single sign-on"}">
      </div>

      ${raw(preset.issuer_template === "{tenant}" || kind !== "oidc" ? html`
        <div class="field">
          <label for="f_tenant">${preset.tenant_label || "Issuer URL"}</label>
          <input id="f_tenant" name="tenant" required
                 value="${existing?.tenant || ""}">
          <div class="hint">${preset.tenant_hint || ""}</div>
        </div>` : "")}

      <div class="field">
        <label for="f_client">Application (client) ID</label>
        <input id="f_client" name="client_id" required
               value="${existing?.client_id || ""}">
      </div>

      <div class="field">
        <label for="f_secret">Client secret</label>
        <input id="f_secret" name="client_secret" type="password"
               autocomplete="new-password"
               placeholder="${existing?.has_secret
                 ? "stored — leave blank to keep it" : ""}">
        <div class="hint">Encrypted before it is written down, and returned by
          no page here, including this one.</div>
      </div>

      <div class="field">
        <label for="f_domains">Only these email domains may sign in</label>
        <input id="f_domains" name="allowed_domains"
               value="${existing?.allowed_domains || ""}"
               placeholder="example.com, example.org">
        <div class="hint">${raw(kind === "google" ? html`<strong>Set this.</strong>
          Google will happily authenticate every Gmail account on earth;
          without a domain here, all of them may sign in.`
          : "Leave empty to accept every address the provider vouches for.")}</div>
      </div>

      <div class="field">
        <label for="f_admins">Groups whose members become administrators</label>
        <input id="f_admins" name="admin_groups"
               value="${existing?.admin_groups || ""}"
               placeholder="AI Studio Admins">
        <div class="hint">Promotion only, never demotion — a group claim that
          is briefly missing must not strip you of your own studio.</div>
      </div>

      <label class="check"><input type="checkbox" name="auto_create"
        ${existing?.auto_create !== false ? "checked" : ""}>
        Create an account the first time somebody signs in</label>
      <label class="check"><input type="checkbox" name="link_by_email"
        ${existing?.link_by_email !== false ? "checked" : ""}>
        Attach to an existing account with the same email address</label>

      ${raw(!preset.directory ? "" : html`
        <h4 style="margin:18px 0 6px">Read the list of people</h4>
        <p class="muted tiny" style="margin:0 0 8px">Imports everybody as a
          findable account so you can share a run with a colleague before they
          have ever opened the studio. Runs hourly.</p>
        <label class="check"><input type="checkbox" name="sync_enabled"
          ${existing?.sync_enabled ? "checked" : ""}>
          Sync from ${preset.directory}</label>
        <div class="field">
          <label for="f_group">${kind === "google"
            ? "Read the directory as (an administrator address)"
            : "Only import members of this group (optional)"}</label>
          <input id="f_group" name="${kind === "google" ? "sync_subject" : "sync_group"}"
                 value="${(kind === "google" ? existing?.sync_subject
                                                : existing?.sync_group) || ""}"
                 placeholder="${kind === "google" ? "admin@example.com"
                   : "AI Studio Users"}">
          <div class="hint">${kind === "google"
            ? "Google's directory can only be read as a real administrator, "
              + "through domain-wide delegation."
            : "Without one, everybody in the tenant is imported."}</div>
        </div>
        <div class="field">
          <label for="f_syncsecret">${kind === "google"
            ? "Service account key (JSON)" : "Directory secret (optional)"}</label>
          ${raw(kind === "google"
            ? html`<textarea id="f_syncsecret" name="sync_secret" rows="4"
                     placeholder="${existing?.has_sync_secret
                       ? "stored — leave blank to keep it"
                       : "paste the whole JSON key file"}"></textarea>`
            : html`<input id="f_syncsecret" name="sync_secret" type="password"
                     autocomplete="new-password"
                     placeholder="${existing?.has_sync_secret
                       ? "stored — leave blank to keep it"
                       : "leave blank to reuse the client secret above"}">`)}
          <div class="hint">${kind === "google"
            ? "Needs the admin.directory.user.readonly scope."
            : "The app registration needs the Graph application permission "
              + "User.Read.All, granted with admin consent."}</div>
        </div>`)}

      <div id="idpError"></div>
      <div class="row" style="gap:8px;margin-top:16px;justify-content:flex-end">
        <button type="button" class="btn" data-modal-close>Cancel</button>
        <button type="submit" class="btn-primary" id="idpSave">
          ${existing ? "Save" : "Connect"}</button>
      </div>
    </form>`;

  on(dlg, "click", "[data-copy-redirect]", () => {
    navigator.clipboard.writeText(redirect)
      .then(() => toast("Copied.", "ok"))
      .catch(() => toast("Could not copy.", "err"));
  });

  $("#idpForm", dlg).addEventListener("submit", async (e) => {
    e.preventDefault();
    const form = e.target;
    const body = { kind };
    $$("input, textarea", form).forEach((el) => {
      if (!el.name) return;
      body[el.name] = el.type === "checkbox" ? el.checked : el.value.trim();
    });
    // An empty secret box on an edit means "keep the one you have", not
    // "delete it". Only a create may send an empty string.
    if (existing) {
      if (!body.client_secret) delete body.client_secret;
      if (!body.sync_secret) delete body.sync_secret;
      if (body.tenant === (existing.tenant || "")) delete body.tenant;
    }
    const btn = $("#idpSave", dlg);
    btn.disabled = true;
    btn.textContent = "Talking to the provider…";
    try {
      if (existing) await api.updateIdp(existing.id, body);
      else await api.createIdp(body);
      dlg.close();
      toast(existing ? "Saved." : "Connected. Try it from a private window.", "ok");
      await refresh();
    } catch (ex) {
      $("#idpError", dlg).innerHTML =
        `<div class="callout callout-err">${esc(ex.message)}</div>`;
      btn.disabled = false;
      btn.textContent = existing ? "Save" : "Connect";
    }
  });
}
